src/clustering/kmeans/kmeans.cc

```diff
@@ -0,0 +1,462 @@
+    }
+    auto closest_centroid_distance = std::make_unique<float[]>(n_train);
+
+    computeClosestCentroid(train_data, n_train, centroids_.get(), closest_centroid, closest_centroid_distance.get());
+    for (size_t i = 0; i < n_train; ++i) {
+        closest_docs[closest_centroid[i]].push_back(i);
+    }
+    std::memset((void*)centroids_.get(), 0x0, num_clusters_ * dim_ * sizeof(DataType));
+    std::vector<int> hassign(num_clusters_, 0);
+
+    std::vector<folly::Future<folly::Unit>> futures;
+    futures.reserve(num_clusters_);
+
+    // compute the new centroids
+    for (size_t i = 0; i < num_clusters_; ++i) {
+        futures.emplace_back(pool_->push([&, c = i]() {
+            hassign[c] = closest_docs[c].size();
+            if (closest_docs[c].empty()) {
+                return;
+            }
+            std::vector<double> centroids_tmp(dim_, 0.0);
+            for (size_t ii = 0; ii < closest_docs[c].size(); ii++) {
+                if (ii + 1 < closest_docs[c].size()) {
+                    auto i1 = closest_docs[c][ii + 1];
+                    // this helps a bit
+                    _mm_prefetch(train_data + i1 * dim_, _MM_HINT_T0);
+                }
+
+                size_t offset = closest_docs[c][ii];
+                for (size_t j = 0; j < dim_; ++j) {
+                    centroids_tmp[j] += double(train_data[offset * dim_ + j]);
+                }
+            }
+            for (size_t j = 0; j < dim_; ++j) {
+                centroids_[c * dim_ + j] = DataType(centroids_tmp[j] / closest_docs[c].size());
+            }
+        }));
+    }
+    for (auto& future : futures) {
+        future.wait();
+    }
+    futures.clear();
+
+    for (size_t i = 0; i < n_train; ++i) {
+        losses += closest_centroid_distance[i];
+    }
+    split_clusters(hassign, n_train, random_state);
+    return losses;
+}
+
+template <typename DataType>
+void
+KmeansClusteringNode<DataType>::computeClosestCentroid(const DataType* vecs, size_t n, const DataType* centroids,
+                                                       uint32_t* closest_centroid, float* closest_centroid_distance) {
+    std::vector<folly::Future<folly::Unit>> futures;
+    constexpr int block_size = 8192;
+    size_t block_num = DIV_ROUND_UP(n, block_size);
+    futures.reserve(block_num);
+    for (size_t i = 0; i < block_num; ++i) {
+        size_t start = i * block_size;
+        size_t end = std::min(n, (i + 1) * block_size);
+        futures.emplace_back(pool_->push([&, start, end]() {
+            if (std::is_same_v<DataType, float>) {
+                exhaustive_L2sqr_blas(vecs + start * dim_, centroids, dim_, end - start, num_clusters_,
+                                      closest_centroid + start, closest_centroid_distance + start);
+            } else {
+                elkan_L2(vecs + start * dim_, centroids, dim_, end - start, num_clusters_, closest_centroid + start,
+                         closest_centroid_distance + start);
+            }
```

Comment:
Looks like this is for float. If we don't support other data types, let's simply return error.

Reply:
remove elkan

Reply:
cluster factory has limited data type to be float, already have a static assert